independent-projects/bootstrap/runner/src/main/java/io/quarkus/bootstrap/runner/JarResource.java

```diff
@@ -28,27 +26,17 @@
 public class JarResource implements ClassLoadingResource {
 
     private final ManifestInfo manifestInfo;
-    private final Path jarPath;
-
-    private final Lock readLock;
-    private final Lock writeLock;
+    final Path jarPath;
 
     private volatile ProtectionDomain protectionDomain;
 
-    //Guarded by the read/write lock; open/close operations on the JarFile require the exclusive lock,
-    //while using an existing open reference can use the shared lock.
-    //If a lock is acquired, and as long as it's owned, we ensure that the zipFile reference
-    //points to an open JarFile instance, and read operations are valid.
-    //To close the jar, the exclusive lock must be owned, and reference will be set to null before releasing it.
-    //Likewise, opening a JarFile requires the exclusive lock.
-    private volatile JarFile zipFile;
+    final AtomicReference<JarFileReference> jarFileReference = new AtomicReference<>();
```

Comment:
I would prefer a single state here. The completable future is good for both cases because allow to be created already completed or can be shared completing it only if we win the race to set it, making easier to follow what's going on.